ctk/common/src/main/kotlin/org/codice/compliance/verification/core/SamlDefinedIdentifiersVerifier.kt

```diff
@@ -118,4 +130,46 @@ internal class SamlDefinedIdentifiersVerifier(val node: Node) {
             }
         }
     }
+
+    /** 8.3.7 Persistent Identifier */
+    private fun verifyPersistentIdentifiers() {
+        node.recursiveChildren()
+            .filter { it.namespaceURI != null }
+            .filter { it.namespaceURI == PERSISTENT_ID }
+            .forEach {
+                if (it.textContent != null && it.textContent.length > ID_VALUE_LENGTH_LIMIT)
+                    throw SAMLComplianceException.create(SAMLCore_8_3_7_a,
+                        message = "The length of the Persistent ID's value was greater than " +
+                            "$ID_VALUE_LENGTH_LIMIT characters.",
+                        node = it)
+
+                if (it.attributeText("NameQualifier") != idpMetadata.entityId )
+                    throw SAMLComplianceException.create(SAMLCore_8_3_7_b,
+                        SAMLCore_8_3_7_c,
+                        message = "The Persistent ID's NameQualifier isn't ${idpMetadata.entityId}",
+                        node = it)
+
+                if (it.attributeText("SPNameQualifier") != SP_ISSUER )
+                    throw SAMLComplianceException.create(SAMLCore_8_3_7_d,
+                        message = "The Persistent ID's SPNameQualifier isn't $SP_ISSUER",
```

Comment:
✏️ You could put the actual value in the exception message. I've used `.let` in the past to reuse the attribute's text value before.
```
it.attributText("SPNameQualifier").let {
        if (it != SP_ISSUER )
                    throw SAMLComplianceException.create(SAMLCore_8_3_7_d,
                        message = "The Persistent ID's SPNameQualifier [$it] isn't equal to $SP_ISSUER",
                        node = it)
}
```